### PR TITLE
Place allfinite tensor on CPU when using Python FE

### PR DESCRIPTION
**Pull request by suffiank**
Split from https://github.com/microsoft/onnxruntime/pull/5354

**Description**: Describe your changes.
Request 'all_finite' output tensor is bound to CPU

**Motivation and Context**
- Why is this change required? What problem does it solve?
The 'all_finite' tensor is bound to GPU by Python FE despite being output to CPU by IsAllFinite kernel. This causes the Python FE to block the CPU until the session is complete. It also causes needless CPU => GPU => CPU transfers.

### CODE REVIEW

**Review by thiagocrepaldi — DISMISSED** [review]

**Comment by thiagocrepaldi** [comment]
This PRs includes submodules updates to cmake/external/SafeInt/safeint. Should it?

File: orttraining/orttraining/python/training/orttrainer.py [hunk 1]
```diff
@@ -787,7 +787,12 @@ def _training_session_run_helper(self, is_train, inputs, inputs_desc, outputs_de
         outputs_desc_resolved = self._resolve_symbolic_dimensions(inputs, inputs_desc, outputs_desc)
         result = {}
         for output_desc in outputs_desc_resolved:
-            torch_tensor = torch.zeros(output_desc.shape, device=self.options.device.id,
+            target_device = self.options.device.id
+            if self.options.mixed_precision.enabled and output_desc.name == self.model_desc.all_finite.name:
```
Comment (KeDengMS) [inline orttraining/orttraining/python/training/orttrainer.py:791]: >self.options.mixed_precision.enabled [](start = 15, length = 36)

Why need to check if mixed_precision is enabled? Seems IsAllFinite covers fp32 as well.

Comment (suffiank) [reply orttraining/orttraining/python/training/orttrainer.py:791]: It seems like `self.model_desc.all_finite.name` is not defined when mixed precision is disabled. How is all_finite used in fp32 case if loss scaling is only applied for mixed precision calculations? Is NaN gradients a particular problem for fp16 cases but not fp32?

Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/orttrainer.py:791]: Current design allows all_finite for mixed_precision only.

**Review by thiagocrepaldi — APPROVED** [review]

**Review by KeDengMS — APPROVED** [review]
:shipit:

**Review by thiagocrepaldi — APPROVED** [review]
